"""
Method 1:
In this primitive approach, once we have the genotype dictionary or motif which are generated by the CRISPRnano which are define by indel location and reference, 
we run the simple Python code to quantify the sequence occurrence directly from fastq file without the need of alignment. 
The drawback of this solution is if the genotype dictionary is too short or in the repetitive region, we may have some false detection result. 
For ONT the frequently single base mutation will affect the result also 
"""
from Bio import SeqIO
def count_specific_sequence_in_fastq(fastq_file, target_sequence):
    specific_sequence_count = 0
    # Iterate loop over sequences in the FASTQ file
    for record in SeqIO.parse(fastq_file, "fastq"):
        # Compare the sequence string with the target sequence
        if target_sequence in str(record.seq):
            specific_sequence_count += 1
    return specific_sequence_count

# Demo usage
fastq_file = "test.fastq"
# This is demo for very short sequence contain insertion TT wrt reference below
target_sequence = "AATTTTCCCA" 
reference =   "AATTCCCA"
num_specific_sequences = count_specific_sequence_in_fastq(fastq_file, target_sequence)

print(f"The specific sequence {target_sequence} occurs {num_specific_sequences} times in the FASTQ file.")
